### PR TITLE
[Feature] Add track_token_ids for Efficient Selective Token Logprobs Tracking

### PR DESCRIPTION
**Pull request by JaviS-Rei**
## Purpose
This PR adds a new `track_token_ids` parameter to `SamplingParams` that enables efficient tracking of log probabilities (logprobs) for specific tokens without requiring full vocabulary logprobs retrieval. This feature addresses [Issue#29280](https://github.com/vllm-project/vllm/issues/29280)

## Implementation Overview

The feature is implemented across the vLLM V1 engine pipeline, from parameter validation to output processing.

### 1. Parameter Extension (`vllm/sampling_params.py`)

Added `track_token_ids: list[int] | None` parameter with validation ensuring all IDs are non-negative integers:

```python
track_token_ids: list[int] | None = None
"""List of token IDs to track logprobs for at every generation step."""
```

### 2. New Data Structures (`vllm/v1/outputs.py`)

Introduced `TrackedLogprobsTensors` (GPU-side) and `TrackedLogprobsLists` (CPU-side for serialization):

- `TrackedLogprobsTensors`: Holds `[batch_size, num_tracked_tokens]` logprobs tensor and token IDs tensor
- `TrackedLogprobsLists`: NumPy/list form for efficient IPC, with `slice_request()` for per-request extraction

### 3. Batch Management (`vllm/v1/worker/gpu_input_batch.py`)

- Tracks `track_token_ids` per request in `InputBatch`
- `merged_track_token_ids` property computes the union of all tracked tokens across the batch as a sorted tensor
- Validates token IDs against vocab bounds before GPU operations

### 4. Sampling Metadata Propagation

Extended `SamplingMetadata` in both:
- `vllm/v1/sample/metadata.py` (standard sampling)
- `vllm/v1/worker/gpu/sample/metadata.py` (GPU worker)

### 5. Core Logprobs Computation

**Standard Sampler (`vllm/v1/sample/sampler.py`):**
```python
@staticmethod
def compute_tracked_logprobs(
    logits_or_logprobs: torch.Tensor,
    track_token_ids: torch.Tensor,
    is_logprobs: bool = False,
) -> TrackedLogprobsTensors:
    """Efficiently extract logprobs for tracked tokens via column indexing."""
    if is_logprobs:
        logprobs = logits_or_logprobs
    else:
        logprobs = logits_or_logprobs.log_softmax(dim=-1, dtype=torch.float32)
    tracked_logprobs = logprobs[:, track_token_ids]  # O(num_tracked)
    return TrackedLogprobsTensors(logprobs=tracked_logprobs, token_ids=track_token_ids)
```

**GPU Worker Sampler (`vllm/v1/worker/gpu/sample/sampler.py`):**
- Calls `compute_tracked_logprobs()` from `vllm/v1/worker/gpu/sample/logprob.py`
- Uses same logits as regular logprobs (raw or processed based on `logprobs_mode`)

**Rejection Sampler for Speculative Decoding (`vllm/v1/sample/rejection_sampler.py`):**
- `_get_tracked_logprobs_tensors()` handles variable-length accepted tokens
- Combines target and bonus logits, extracts tracked logprobs only for accepted positions

### 6. Output Processing Pipeline

**Model Runner (`vllm/v1/worker/gpu_model_runner.py`):**
- Extracts `tracked_logprobs_tensors` from `SamplerOutput`
- Converts to `TrackedLogprobsLists` via `tolists()` with `cu_num_tokens` for proper slicing
- Handles async scheduling with `AsyncGPUModelRunnerOutput`

**Scheduler (`vllm/v1/core/sched/scheduler.py`):**
- Slices tracked logprobs per request using `slice_request()`
- Attaches `new_tracked_logprobs` to `EngineCoreOutput`

**Logprobs Processor (`vllm/v1/engine/logprobs.py`):**
- Accumulates tracked logprobs across generation steps in `LogprobsProcessor`
- Handles merged token IDs from batch by mapping back to per-request `track_token_ids`

**Output Processor (`vllm/v1/engine/output_processor.py`):**
- Includes `tracked_logprobs` in `CompletionOutput`
- Supports DELTA mode (returns only new tokens' logprobs)

### 7. Output Extension (`vllm/outputs.py`)

Added `tracked_logprobs: dict[int, list[float]] | None` to `CompletionOutput` with merge logic in `RequestOutput` for streaming.

---

## Test Design

The test suite provides comprehensive coverage at multiple levels:

### 1. Unit Tests - Parameter Validation (`tests/v1/test_sampling_params.py`)

- Valid inputs: `None`, empty list `[]`, single/multiple token IDs, boundary values (0, large values)
- Invalid inputs: negative integers raise `ValueError`
- Coexistence: works with `logprobs`, `prompt_logprobs`, `temperature`

### 2. Unit Tests - Core Computation (`tests/v1/sample/test_sampler.py`)

- `test_compute_tracked_logprobs_from_logits`: Verifies correct shape and values with manual log_softmax
- `test_compute_tracked_logprobs_from_logprobs`: Verifies `is_logprobs=True` skips redundant log_softmax
- Parametrized across devices, batch sizes, and number of tracked tokens
- Edge cases: single token, boundary tokens (0 and vocab_size-1), many tokens (100)

### 3. Unit Tests - Data Structures (`tests/v1/test_outputs.py`)

- `TrackedLogprobsLists.slice_request()` correctness with/without `cu_num_generated_tokens`
- Slicing for first/middle/last requests, single token tracking, many tokens (100)

### 4. Unit Tests - Batch Management (`tests/v1/worker/test_gpu_input_batch.py`)

- `merged_track_token_ids` correctly unions and sorts across requests
- Empty cases return `None`
- Vocab bounds validation raises on out-of-range token IDs

### 5. Unit Tests - Output Processing (`tests/v1/engine/test_output_processor.py`)

- `test_tracked_logprobs_basic`: Full pipeline with DELTA and FINAL_ONLY modes
- `test_tracked_logprobs_none_when_not_requested`: Returns `None` when disabled
- `test_tracked_logprobs_with_sample_logprobs`: Coexists with regular logprobs

### 6. End-to-End Tests (`tests/v1/e2e/test_tracked_logprobs.py`) - **1,656 lines**

**Basic Functionality:**
- `test_track_token_ids_returns_logprobs`: Verifies logprobs returned for all tracked tokens
- `test_track_token_ids_with_logprobs_coexist`: Both regular and tracked logprobs work together

**None vs Empty:**
- `test_track_token_ids_not_provided` / `test_track_token_ids_explicit_none`: Returns `None`
- `test_track_token_ids_empty_list`: Returns empty dict `{}`

**Batch Processing:**
- Multiple prompts with same track_token_ids
- Different completion lengths

**Differential Tests (Critical):**
- **Uses BOTH `logprobs=-1` AND `track_token_ids` in the same request**
- Compares `tracked_logprobs[tid]` vs `full_logprobs[step][tid].logprob`
- Parametrized across sampling configurations (greedy, temperature, top_p, top_k)
- Ensures values match exactly since they come from the same computation

**Streaming Tests (AsyncLLM):**
- `test_streaming_basic`: DELTA mode accumulates correctly
- `test_streaming_track_token_ids_none`: Returns `None` in streaming
- `test_streaming_vs_nonstreaming_consistency`: Final accumulated values match non-streaming

---

## Documentation

- Added `docs/features/track_token_ids.md` with motivation, use cases, API reference, and examples
- Updated `docs/usage/v1_guide.md` to reference the new feature

---

## Backward Compatibility

✅ **Fully backward compatible:**
- `track_token_ids=None` (default): No behavior change
- Existing code continues to work unchanged
- Feature is opt-in

### CODE REVIEW

**Comment by zza234s** [comment]
Thanks for your contribution and the great work !!!

I noticed this PR currently has some merge conflicts and a failing check. Do you have plans to address them, or is this PR already ready for experimental use? I'd love to try out this feature in my project.

**Comment by JaviS-Rei** [comment]
@zza234s Merge conflicts are resolved now. This PR is already ready for experimental use. Welcome to try this feature and report any questions or bugs!

**Comment by JaviS-Rei** [comment]
@zza234s This PR is currently under test and merged since the main branch updated. Hopefully, it will be available on Monday.

**Comment by JaviS-Rei** [comment]
@mergifyio refresh

**Comment by JaviS-Rei** [comment]
@zza234s This PR is currently tested. Feel free to report any questions or bugs you encounter.

**Comment by zza234s** [comment]
> @zza234s This PR is currently tested. Feel free to report any questions or bugs you encounter.

Thanks a lot for your hard work. Much appreciated ❤️

**Comment by JaviS-Rei** [comment]
@cboss6 This PR is now paused, or it doesn't catch up with upstream. The current version (my fork repo) is tested. If you need new feature in upstream,  you may solve conflicts and contribute. Thanks.
